File: DataTransfer/server/calibration.cpp
#include <iostream>
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>
#include <vector>
#include <complex>
#include <cstdint>
#include <cstring>
#include <algorithm>

const char* SERIAL_PORT = "/dev/ttyACM0";
const int BAUDRATE = B115200;
const int POINTS = 201;
const int BLOCK_SIZE = 32;
const uint64_t START_FREQ = 500000000;
const uint64_t STOP_FREQ = 3000000000;
const uint64_t STEP_FREQ = (STOP_FREQ - START_FREQ) / (POINTS - 1);

int openSerial(const char* device) {
    int fd = open(device, O_RDWR | O_NOCTTY);
    if (fd < 0) return -1;

    struct termios options{};
    tcgetattr(fd, &options);
    cfsetspeed(&options, BAUDRATE);
    options.c_cflag |= (CLOCAL | CREAD);
    options.c_cflag &= ~CSIZE;
    options.c_cflag |= CS8;
    options.c_cflag &= ~PARENB;
    options.c_cflag &= ~CSTOPB;
    options.c_cflag &= ~CRTSCTS;
    options.c_iflag = IGNPAR;
    tcflush(fd, TCIFLUSH);
    tcsetattr(fd, TCSANOW, &options);
    return fd;
}

void write_reg_8(int fd, uint8_t addr, uint64_t value) {
    uint8_t buf[10];
    buf[0] = 0x23;
    buf[1] = addr;
    for (int i = 0; i < 8; ++i)
        buf[2 + i] = (value >> (i * 8)) & 0xFF;
    write(fd, buf, 10);
}

void write_reg_2(int fd, uint8_t addr, uint16_t value) {
    uint8_t buf[4] = {0x21, addr, (uint8_t)(value & 0xFF), (uint8_t)((value >> 8) & 0xFF)};
    write(fd, buf, 4);
}

void flush_fifo(int fd) {
    uint8_t cmd[2] = {0x30, 0};
    write(fd, cmd, 2);
    usleep(50000);
    write(fd, cmd, 2);
    usleep(50000);
}

void trigger_sweep(int fd) {
    uint8_t cmd[2] = {0x27, 1};
    write(fd, cmd, 2);
}

std::vector<uint8_t> readSweep(int fd, int total_bytes) {
    std::vector<uint8_t> buffer(total_bytes);
    int bytes_read = 0;
    while (bytes_read < total_bytes) {
        int r = read(fd, buffer.data() + bytes_read, total_bytes - bytes_read);
        if (r <= 0) break;
        bytes_read += r;
    }
    buffer.resize(bytes_read);
    return buffer;
}

void parseAndPrint(const std::vector<uint8_t>& data) {
    int total = std::min((int)data.size() / BLOCK_SIZE, POINTS);
    std::cout << "Total valid blocks: " << total << std::endl;
    for (int i = 0; i < total; ++i) {
        const uint8_t* block = &data[i * BLOCK_SIZE];
        int32_t fwd_re = *reinterpret_cast<const int32_t*>(&block[0]);
        int32_t fwd_im = *reinterpret_cast<const int32_t*>(&block[4]);
        int32_t rev_re = *reinterpret_cast<const int32_t*>(&block[8]);
        int32_t rev_im = *reinterpret_cast<const int32_t*>(&block[12]);
        uint16_t idx = *reinterpret_cast<const uint16_t*>(&block[24]);
        std::complex<float> fwd(fwd_re, fwd_im);
        std::complex<float> rev(rev_re, rev_im);
        std::complex<float> s11 = (std::abs(fwd) > 0) ? rev / fwd : std::complex<float>(0, 0);
        double freq = START_FREQ + idx * STEP_FREQ;
        std::cout << freq / 1e6 << " MHz: " << s11.real() << " + j" << s11.imag() << std::endl;
    }
}

void sweepAndShow(int fd, const char* label) {
    std::string prompt;
    std::cout << "Connect " << label << " to CH0, then press ENTER..." << std::endl;
    std::getline(std::cin, prompt);
    flush_fifo(fd);
    trigger_sweep(fd);
    usleep(400000);
    auto data = readSweep(fd, POINTS * BLOCK_SIZE);
    std::cout << "\n[" << label << "] Sweep Result:" << std::endl;
    parseAndPrint(data);
}

int main() {
    int fd = openSerial(SERIAL_PORT);
    if (fd < 0) {
        std::cerr << "Error opening serial port." << std::endl;
        return 1;
    }

    write_reg_8(fd, 0x00, START_FREQ);
    write_reg_8(fd, 0x10, STEP_FREQ);
    write_reg_2(fd, 0x20, POINTS);

    sweepAndShow(fd, "LOAD");
    sweepAndShow(fd, "OPEN");
    sweepAndShow(fd, "SHORT");

    close(fd);
    return 0;
}

File: DataTransfer/server/calibration.py
import serial
import struct
import time

PORT = "/dev/ttyACM0"
BAUDRATE = 115200
POINTS = 201
START_FREQ = 500_000_000
STOP_FREQ = 3_000_000_000
STEP_FREQ = int((STOP_FREQ - START_FREQ) / (POINTS - 1))

def write_reg_8(ser, addr, value):
    ser.write(bytes([0x23, addr]) + value.to_bytes(8, 'little'))

def write_reg_2(ser, addr, value):
    ser.write(bytes([0x21, addr, value & 0xFF, (value >> 8) & 0xFF]))

def trigger_sweep(ser):
    ser.write(bytes([0x27, 1]))

def flush_fifo(ser):
    ser.write(bytes([0x30, 0]))
    time.sleep(0.05)
    ser.write(bytes([0x30, 0]))
    time.sleep(0.05)

def read_sweep(ser, points):
    ser.write(bytes([0x18, 0x30, points]))
    data = ser.read(32 * points)
    return data

def parse_sweep(data, points):
    results = []
    for i in range(points):
        block = data[i*32:(i+1)*32]
        if len(block) < 32:
            continue
        rev_re = struct.unpack('<i', block[8:12])[0]
        rev_im = struct.unpack('<i', block[12:16])[0]
        fwd_re = struct.unpack('<i', block[0:4])[0]
        fwd_im = struct.unpack('<i', block[4:8])[0]
        freq_index = struct.unpack('<H', block[24:26])[0]
        freq = START_FREQ + freq_index * STEP_FREQ
        fwd = complex(fwd_re, fwd_im)
        rev = complex(rev_re, rev_im)
        s11 = rev / fwd if fwd != 0 else 0
        results.append((freq / 1e6, s11))
    return sorted(results, key=lambda x: x[0])

def sweep_and_print(label):
    input(f"Connect {label} to CH0, then press ENTER...")
    flush_fifo(ser)
    trigger_sweep(ser)
    time.sleep(0.4)
    data = read_sweep(ser, POINTS)
    samples = parse_sweep(data, POINTS)
    print(f"\n[{label}] Sweep Results: {len(samples)} samples (sorted)")
    for f, s in samples:
        print(f"{f:9.3f} MHz | {s.real:+.3e}{s.imag:+.3e}j")
    return samples

with serial.Serial(PORT, BAUDRATE, timeout=1) as ser:
    write_reg_8(ser, 0x00, START_FREQ)
    write_reg_8(ser, 0x10, STEP_FREQ)
    write_reg_2(ser, 0x20, POINTS)

    load_data = sweep_and_print("LOAD")
    open_data = sweep_and_print("OPEN")
    short_data = sweep_and_print("SHORT")

